File: api/routers/transfers.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy import or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from ..database import get_db
from ..dependencies import current_user
from ..models.transfer import Transfer
from ..models.user import User
from ..schemas.transfers import TransferCreate, TransferResponse
from ..services.transfers import create_transfer_with_transactions

router = APIRouter(prefix="/api/v1/transfers", tags=["transfers"])


@router.post("", response_model=TransferResponse, status_code=201)
async def create_transfer(
    payload: TransferCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(current_user),
):
    result = await create_transfer_with_transactions(
        db, user_id=user.id, payload=payload
    )
    await db.commit()
    await db.refresh(result.transfer)
    response = TransferResponse.model_validate(result.transfer)
    return response.model_copy(
        update={
            "debit_transaction_id": result.debit_transaction_id,
            "credit_transaction_id": result.credit_transaction_id,
        }
    )


@router.get("", response_model=list[TransferResponse])
async def list_transfers(
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    account_id: uuid.UUID | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(current_user),
):
    stmt = select(Transfer).where(Transfer.user_id == user.id)
    if account_id is not None:
        # Either side — the native account detail lists every transfer that
        # touches the account (Phase 7b).
        stmt = stmt.where(
            or_(
                Transfer.from_account_id == account_id,
                Transfer.to_account_id == account_id,
            )
        )
    result = await db.execute(
        stmt.order_by(Transfer.occurred_at.desc(), Transfer.created_at.desc())
        .offset(offset)
        .limit(limit)
    )
    return list(result.scalars().all())

File: mobile/src/api/transfers.ts
/**
 * Phase 7b B1 — transfers between own accounts.
 *
 * A credit-card payment IS a transfer (source account → credit account): the
 * two legs share a transfer_id and are excluded from income/expense math, so
 * paying the card never double-counts spending.
 */
import { api } from "./client";

export interface TransferCreate {
  from_account_id: string;
  to_account_id: string;
  amount: string;
  currency: string;
  fx_rate?: string | null;
  occurred_at?: string | null;
  notes?: string | null;
}

export interface TransferResponse {
  id: string;
  user_id: string;
  from_account_id: string;
  to_account_id: string;
  amount: string;
  currency: string;
  fx_rate: string | null;
  occurred_at: string;
  notes: string | null;
  created_at: string;
  debit_transaction_id: string | null;
  credit_transaction_id: string | null;
}

export async function createTransfer(
  payload: TransferCreate,
): Promise<TransferResponse> {
  const { data } = await api.post<TransferResponse>("/transfers", payload);
  return data;
}

export async function fetchTransfers(
  accountId?: string,
  limit = 50,
  offset = 0,
): Promise<TransferResponse[]> {
  const { data } = await api.get<TransferResponse[]>("/transfers", {
    params: {
      limit,
      offset,
      ...(accountId ? { account_id: accountId } : {}),
    },
  });
  return data;
}
